src/main/java/dk/bec/polonez/reservationsystem/service/OfferService.java

```diff
@@ -35,32 +42,76 @@ public OfferService(OfferRepository offerRepository, FeatureRepository featureRe
+        Optional<Offer> optionalOffer = offerRepository.findById(id);
+        Offer offer = optionalOffer.orElseThrow(
+                () -> new NotFoundObjectException(Offer.class, id));
+
+        User currentUser = authService.getCurrentUser();
+        Role currentUserRole = currentUser.getRole();
+
+        if (!currentUserRole.hasOfferUpdatePrivilege())
+            throw new NoAccessToOfferOperationException();
+
+        if (!offer.getOwner().equals(currentUser) && !currentUserRole.hasOfferUpdateOthersOfferPrivilege())
+            throw new NoAccessToOfferOperationException("You cannot update other Place Owner's offer as a Place Owner, to do that log as an Admin");
+
         modelMapper.map(offerDto, offer);
         Offer updatedOffer = offerRepository.save(offer);
         return modelMapper.map(updatedOffer, OfferDto.class);
     }
 
     public OfferDto deleteOffer(long id) {
-        Offer offerToDelete = offerRepository.getById(id);
-        offerRepository.delete(offerToDelete);
-        return modelMapper.map(offerToDelete, OfferDto.class);
+        Optional<Offer> optionalOffer = offerRepository.findById(id);
+        Offer offer = optionalOffer.orElseThrow(
+                () -> new NotFoundObjectException(Offer.class, id));
+
+        User currentUser = authService.getCurrentUser();
+        Role currentUserRole = currentUser.getRole();
+
+        if (!currentUserRole.hasOfferDeletePrivilege())
+            throw new NoAccessToOfferOperationException();
+        if (!offer.getOwner().equals(currentUser) && !currentUserRole.hasOfferDeleteOthersOfferPrivilege())
+            throw new NoAccessToOfferOperationException("You cannot delete other Place Owner's offer as a Place Owner, to do that log as an Admin");
+
+        offerRepository.delete(offer);
+        return modelMapper.map(offer, OfferDto.class);
     }
 
     public ResponseOfferFeatureDto addFeatureToOffer(long featureId, long offerId) {
-        Feature feature = featureRepository.getById(featureId);
-        Offer offer = offerRepository.getById(offerId);
+        Optional<Feature> optionalFeature = featureRepository.findById(featureId);
+        Optional<Offer> optionalOffer = offerRepository.findById(offerId);
+        Offer offer = optionalOffer.orElseThrow(
+                () -> new NotFoundObjectException(Offer.class, offerId));
+        Feature feature = optionalFeature.orElseThrow(
+                () -> new NotFoundObjectException(Feature.class, featureId));
+
+        User currentUser = authService.getCurrentUser();
+        Role currentUserRole = currentUser.getRole();
+
+        if (!currentUserRole.hasOfferUpdatePrivilege())
+            throw new NoAccessToOfferOperationException();
+        if (!offer.getOwner().equals(currentUser) && !currentUserRole.hasOfferDeleteOthersOfferPrivilege)
+            throw new NoAccessToOfferOperationException("You cannot update other Place Owner's offer features as a Place Owner, to do that log as an Admin");
```

Comment:
**advice** Exception class name should explain what this exception is about business-wise - and then a ```String``` message (with explanation what happened) will become redundant and you will be able to remove it.

Generally speaking, if you're re-using single exception but you throw it in different places in the source code with different ```String``` messages, then you can consider defining seprate exception for each case (= no reuse). Each separate exception would have single responsibility